mkosi/machine.py

```diff
@@ -0,0 +1,125 @@
+# SPDX-License-Identifier: LGPL-2.1+
+
+from __future__ import annotations
+
+import contextlib
+import signal
+import subprocess
+from textwrap import dedent
+from typing import Any, Optional, Sequence
+
+import pexpect  # type: ignore
+
+from . import (
+    MKOSI_COMMANDS_SUDO,
+    CompletedProcess,
+    build_stuff,
+    check_native,
+    check_output,
+    check_root,
+    init_namespace,
+    load_args,
+    needs_build,
+    parse_args,
+    prepend_to_environ_path,
+    run_command_image,
+    run_qemu_cmdline,
+    run_shell_cmdline,
+    unlink_output,
+)
+from .backend import MkosiArgs, die
+
+
+class Machine:
+    def __init__(self, args: Optional[Sequence[str]] = None, debug: bool = False, test: bool = False) -> None:
+        # Remains None until image is built and booted, then receives pexpect process
+        self._serial: Optional[pexpect.spawn] = None
+        self.exit_code: int = -1
+        self.debug = debug
+        self.stack = contextlib.ExitStack()
+        self.args: MkosiArgs
+
+        # We make sure to add the arguments in the machine class itself, rather than typing this for every testing function.
+        if test:
```

Comment:
Not really, this just replace the hard coded logid from here: https://github.com/systemd/mkosi/pull/894#discussion_r803608343

But maybe this should be removed altogether, if someone needs to type parameters multiple times, just use _parametrize_ from pytest

Reply:
I think we're misunderstanding each other here.  Adding the necessary arguments depending on what the verb is in the constructor is a good thing. What I don't understand is why we gate this behind the `test` boolean argument. Right now, in all your tests, you specify `test=True`. What I don't understand is: in which scenario would users specify `test=False`?

Reply:
That is gated behind `test` because I understood we would only autocomplete arguments to generate a fixed set of images (i.e the ones used in testing currently). So for every other scenario outside of the ones existing on test_machine.py all arguments would actually be specified through `args`.

Reply:
But doing so would make the Machine class quite error prone to use no? Everyone that tried to use it for the first time might forget to add the --ssh option for example and suddenly Machine.run() won't work correctly because there's no SSH inside the virtual machine.

So I'd argue that we definitely by default want to add these arguments.

In the future, there could be special use cases for very specific tests where we don't want to add these arguments, but when that happens we should add an option to disable the arguments from being added, instead of adding an option that has to be enabled to add the extra arguments. This way we'd do the right thing by default but still give users an escape hatch if they need it.

Let's wait with adding that option until we have a concrete use case though. For now I'd go ahead and just remove the `test` argument and always add the extra arguments.